src/sentry/sentry_metrics/indexer/limiters/writes.py

```diff
@@ -109,80 +54,157 @@ def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
-            )
+        if exc_type is not None:
+            return
+        for requests, grants, timestamp in zip(self._requests, self._grants, self._timestamps):
+            self._writes_limiter.rate_limiter.use_quotas(requests, grants, timestamp)
 
 
 class WritesLimiter:
     def __init__(self, namespace: str, **options: Mapping[str, str]) -> None:
         self.namespace = namespace
         self.rate_limiter: RedisSlidingWindowRateLimiter = RedisSlidingWindowRateLimiter(**options)
 
+    def _build_quota_key(self, org_id: Optional[OrgId] = None) -> str:
+        if org_id is not None:
+            return f"metrics-indexer-{self.namespace}-org-{org_id}"
+        else:
+            return f"metrics-indexer-{self.namespace}-global"
+
+    @metrics.wraps("sentry_metrics.indexer.construct_quotas")
+    def _construct_quotas(self, use_case_id: UseCaseID) -> Sequence[Quota]:
+        """
+        Construct write limit's quotas based on current sentry options.
+
+        This value can potentially cached globally as long as it is invalidated
+        when sentry.options are.
+        """
+        if use_case_id in USE_CASE_ID_WRITES_LIMIT_QUOTA_OPTION_NAME:
+            return [
+                Quota(prefix_override=self._build_quota_key(), **args)
+                for args in options.get(
+                    f"{USE_CASE_ID_WRITES_LIMIT_QUOTA_OPTION_NAME[use_case_id]}.global"
+                )
+            ] + [
+                Quota(prefix_override=None, **args)
+                for args in options.get(
+                    f"{USE_CASE_ID_WRITES_LIMIT_QUOTA_OPTION_NAME[use_case_id]}.per-org"
+                )
+            ]
+        raise ValueError(use_case_id)
+
+    @metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+    def _construct_quota_requests(
+        self, use_case_id: UseCaseID, keys: KeyCollection
+    ) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
+        org_ids = []
+        requests = []
+        quotas = self._construct_quotas(use_case_id)
+
+        if quotas:
+            for org_id, strings in keys.mapping.items():
+                org_ids.append(org_id)
+                requests.append(
+                    RequestedQuota(
+                        prefix=self._build_quota_key(org_id),
+                        requested=len(strings),
+                        quotas=quotas,
+                    )
+                )
+
+        return org_ids, requests
+
+    def _check_use_case_writes_limits(
+        self, use_case_id: UseCaseID, key_collection: KeyCollection
+    ) -> Tuple[
+        Timestamp,
+        Sequence[RequestedQuota],
+        Sequence[GrantedQuota],
+        Sequence[DroppedString],
+        dict[OrgId, Set[str]],
+    ]:
+        org_ids, requests = self._construct_quota_requests(
+            use_case_id,
+            key_collection,
+        )
+        timestamp, grants = self.rate_limiter.check_within_quotas(requests)
```

Comment:
i think this function should be called once per batch. otherwise performance will suffer too much

Reply:
Do you mind elaborating on this? How much performance are we loosing? Is it because it's communicating with the redis cluster? Currently we only have 2 use cases and we don't expect the number of use case to be too high.

Reply:
`check_within_quotas` is internally batching its operations to redis. you are adding at least a few milliseconds per batch per added usecase. it might be fine for a few usecases but it will not scale, and it will be a performance regression in the pipeline today

Reply:
I see, thanks. Another quick question: why would the number of calls to `check_within_quotas` per batch matter if `check_within_quotas` is internally batching it's operations to redis?

Reply:
> why would the number of calls to check_within_quotas per batch matter if check_within_quotas is internally batching it's operations to redis?

what i mean is that multiple QuotaRequest objects are batched together in `check_within_quotas`. multiple calls to `check_within_quotas` are not batched together in any way

Reply:
Got it, thanks!